Add CBlockIndex lock annotations, guard nStatus/nFile/nDataPos/nUndoPos by cs_main

jonatack opened the pull request:
Issues:

- `CBlockIndex` member functions `GetBlockPos()`, `GetUndoPos()`, `IsAssumedValid()`, `RaiseValidity()`, and `IsValid()` and block storage functions `WriteUndoDataForBlock()` and `IsBlockPruned()` are missing thread safety lock annotations to help ensure that they are called with mutex cs_main to avoid bugs like #22895. Doing this also enables the next step:

- `CBlockIndex::nStatus` may be racy, i.e. potentially accessed by multiple threads, see #17161. A solution is to guard it by cs_main, along with fellow data members `nFile`, `nDataPos` and `nUndoPos`.

This pull:

- adds thread safety lock annotations for the functions listed above
- guards `CBlockIndex::nStatus`, `nFile`, `nDataPos` and `nUndoPos` by cs_main

How to review and test:
- debug build with clang and verify there are no `-Wthread-safety-analysis` warnings
- review the code to verify each annotation or lock is necessary and sensible, or if any are missing
- look for whether taking a lock can be replaced by a lock annotation instead
- for more information about Clang thread safety analysis, see
    - https://clang.llvm.org/docs/ThreadSafetyAnalysis.html
    - https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md#lockingmutex-usage-notes
    - https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md#threads-and-synchronization

Mitigates/potentially closes #17161.

jonatack committed:
Require CBlockIndex::GetBlockPos() to hold mutex cs_main

jonatack committed:
Require CBlockIndex::GetUndoPos() to hold mutex cs_main

jonatack commented:
Tests green. Added a commit to require `CBlockIndex::GetUndoPos()` to hold cs_main as well, if reviewers think it's a good idea.

jonatack commented:
Rebased following the merge of #22895 that this was based on.

MarcoFalke commented on src/chain.h:
```diff
@@ -212,7 +215,8 @@ class CBlockIndex
     }
 
-    FlatFilePos GetBlockPos() const {
+    FlatFilePos GetBlockPos() const EXCLUSIVE_LOCKS_REQUIRED(cs_main)
```
This is annotating the function, but not the underlying data member that needs the annotation.

Not sure how involved it will be to annotate everything properly here. Also, given that cs_main is the validation lock, we should (eventually/ideally) have a separate node_storage lock. Though, this again requires a major validation/node_storage rewrite.

However an argument against adding the lock annotation to the data member could be that `chain` is supposed to be a data-structure-only primitive, like `transaction` or `block`. In that case only annotating the getter and setter functions (and making them standalone) would seem appropriate.

jonatack replied:
Having a quick look at annotating `nStatus` to assess.

jonatack replied:
Appended an exploratory small commit guarding `CBlockIndex::nStatus` by `cs_main`.

maflcko reviewed (commented):
Concept ACK. Feel free to ignore the meta comment.

MarcoFalke commented on src/test/util/blockfilter.cpp:
```diff
@@ -12,7 +12,7 @@
 {
     CBlock block;
-    if (!ReadBlockFromDisk(block, block_index->GetBlockPos(), Params().GetConsensus())) {
+    if (!ReadBlockFromDisk(block, WITH_LOCK(cs_main, return block_index->GetBlockPos()), Params().GetConsensus())) {
```
nit: unit tests are single-threaded, so you can just take the lock for the whole function. It is likely that there are other lock violations anyway: See https://github.com/bitcoin/bitcoin/issues/17161#issuecomment-543151879

jonatack replied:
done

MarcoFalke commented on src/wallet/test/wallet_tests.cpp:
```diff
@@ -86,7 +86,8 @@ BOOST_FIXTURE_TEST_CASE(scan_for_wallet_transactions, TestChain100Setup)
     CBlockIndex* oldTip = m_node.chainman->ActiveChain().Tip();
-    GetBlockFileInfo(oldTip->GetBlockPos().nFile)->nSize = MAX_BLOCKFILE_SIZE;
+    const auto pos{WITH_LOCK(cs_main, return oldTip->GetBlockPos())};
+    GetBlockFileInfo(pos.nFile)->nSize = MAX_BLOCKFILE_SIZE;
```
Same: Can include this in the cs_main scope as well, I think.

jonatack replied:
done, thank you

jonatack commented:
Thanks @MarcoFalke, updated with your suggestions per `git diff f6c2fdb e950f0e`

<details><summary>diff</summary><p>

```diff 
diff --git a/src/test/util/blockfilter.cpp b/src/test/util/blockfilter.cpp
index 311913b5db..1a663bd0b3 100644
--- a/src/test/util/blockfilter.cpp
+++ b/src/test/util/blockfilter.cpp
@@ -11,8 +11,10 @@
 
 bool ComputeFilter(BlockFilterType filter_type, const CBlockIndex* block_index, BlockFilter& filter)
 {
+    LOCK(cs_main);
+
     CBlock block;
-    if (!ReadBlockFromDisk(block, WITH_LOCK(cs_main, return block_index->GetBlockPos()), Params().GetConsensus())) {
+    if (!ReadBlockFromDisk(block, block_index->GetBlockPos(), Params().GetConsensus())) {
         return false;
     }
 
diff --git a/src/wallet/test/wallet_tests.cpp b/src/wallet/test/wallet_tests.cpp
index 37fd52a5e7..4ac8936ee7 100644
--- a/src/wallet/test/wallet_tests.cpp
+++ b/src/wallet/test/wallet_tests.cpp
@@ -86,8 +86,10 @@ BOOST_FIXTURE_TEST_CASE(scan_for_wallet_transactions, TestChain100Setup)
 {
     // Cap last block file size, and mine new block in a new block file.
     CBlockIndex* oldTip = m_node.chainman->ActiveChain().Tip();
-    const auto pos{WITH_LOCK(cs_main, return oldTip->GetBlockPos())};
-    GetBlockFileInfo(pos.nFile)->nSize = MAX_BLOCKFILE_SIZE;
+    {
+        LOCK(cs_main);
+        GetBlockFileInfo(oldTip->GetBlockPos().nFile)->nSize = MAX_BLOCKFILE_SIZE;
+    }
     CreateAndProcessBlock({}, GetScriptForRawPubKey(coinbaseKey.GetPubKey()));
     CBlockIndex* newTip = m_node.chainman->ActiveChain().Tip();
```
</p></details>

jonatack commented:
Rebased after the merge of #21526.

promag reviewed (commented):
Concept ACK.

jonatack committed:
Require WriteUndoDataForBlock() to hold mutex cs_main

Mutex cs_main is already held by the caller of WriteUndoDataForBlock().
This change is needed to require CBlockIndex::GetUndoPos() to hold
cs_main and CBlockIndex::nStatus to be guarded by cs_main in the
following commits without adding 2 unnecessary cs_main locks to
WriteUndoDataForBlock().

jonatack commented on src/chain.h:
```diff
@@ -220,7 +223,8 @@ class CBlockIndex
 
-    FlatFilePos GetBlockPos() const {
+    FlatFilePos GetBlockPos() const EXCLUSIVE_LOCKS_REQUIRED(cs_main)
+    {
```
Thanks @promag, done.

jonatack commented on src/chain.h:
```diff
@@ -233,7 +233,8 @@ class CBlockIndex
 
-    FlatFilePos GetUndoPos() const {
+    FlatFilePos GetUndoPos() const EXCLUSIVE_LOCKS_REQUIRED(cs_main)
+    {
```
Done.

jonatack commented on src/rpc/rawtransaction.cpp:
```diff
@@ -196,7 +196,8 @@ static RPCHelpMan getrawtransaction()
         std::string errmsg;
         if (blockindex) {
-            if (!(blockindex->nStatus & BLOCK_HAVE_DATA)) {
+            const auto status{WITH_LOCK(cs_main, return blockindex->nStatus)};
```
Done.

jonatack commented on src/node/blockstorage.cpp:
```diff
@@ -336,7 +337,7 @@ static bool WriteBlockToDisk(const CBlock& block, FlatFilePos& pos, const CMessa
 {
     // Write undo information to disk
-    if (pindex->GetUndoPos().IsNull()) {
+    if (WITH_LOCK(cs_main, return pindex->GetUndoPos()).IsNull()) {
```
@promag You're right; Clang 13 seems unable to detect this on its own, but the unneeded lock can be avoided by requiring `WriteUndoDataForBlock()` to hold cs_main instead. Done in  686ae2d97.

jonatack commented on src/node/blockstorage.cpp:
```diff
@@ -355,9 +356,12 @@ bool WriteUndoDataForBlock(const CBlockUndo& blockundo, BlockValidationState& st
-        pindex->nStatus |= BLOCK_HAVE_UNDO;
-        setDirtyBlockIndex.insert(pindex);
+        {
+            LOCK(cs_main);
```
Thanks! Removed in both places in this function as described in https://github.com/bitcoin/bitcoin/pull/22932#discussion_r753832096.

jonatack commented:
Thank you for reviewing, @promag. Updated based on your feedback per `git diff 9cacc50 b1c859a`.

fjahr commented:
Hm, the description only says "Potentially related to issue #17161.", otherwise I don't really see a description of what problem this PR is actually fixing/improving?

laanwj commented:
> otherwise I don't really see a description of what problem this PR is actually fixing/improving?

Is the `nStatus` field potentially accessed by multiple threads at once? #17161 seems to indicate it is. If so, it needs to be either an atomic value or protected by a mutex.

jonatack commented:
Rebased.  @fjahr I'll improve the OP.  @laanwj: thanks!

fanquake commented:
While you're making changes, can you also update the commits to drop the `consensus:` prefix, as I don't think that's correct for these changes.

hebasto commented:
Concept ACK.

hebasto commented on src/node/blockstorage.cpp:
```diff
@@ -445,12 +448,7 @@ bool ReadRawBlockFromDisk(std::vector<uint8_t>& block, const FlatFilePos& pos, c
-    FlatFilePos block_pos;
-    {
-        LOCK(cs_main);
-        block_pos = pindex->GetBlockPos();
-    }
-
+    const FlatFilePos block_pos{WITH_LOCK(cs_main, return pindex->GetBlockPos())};
```
How about getting rid of this function entirely:
```diff
diff --git a/src/net_processing.cpp b/src/net_processing.cpp
index 2a30afdb5..030a8c736 100644
--- a/src/net_processing.cpp
+++ b/src/net_processing.cpp
@@ -1848,7 +1848,7 @@ void PeerManagerImpl::ProcessGetBlockData(CNode& pfrom, Peer& peer, const CInv&
         // Fast-path: in this case it is possible to serve the block directly from disk,
         // as the network format matches the format on disk
         std::vector<uint8_t> block_data;
-        if (!ReadRawBlockFromDisk(block_data, pindex, m_chainparams.MessageStart())) {
+        if (!ReadRawBlockFromDisk(block_data, pindex->GetBlockPos(), m_chainparams.MessageStart())) {
             assert(!"cannot load block from disk");
         }
         m_connman.PushMessage(&pfrom, msgMaker.Make(NetMsgType::BLOCK, Span{block_data}));
diff --git a/src/node/blockstorage.cpp b/src/node/blockstorage.cpp
index 1af24ecf6..bf47ab351 100644
--- a/src/node/blockstorage.cpp
+++ b/src/node/blockstorage.cpp
@@ -446,12 +446,6 @@ bool ReadRawBlockFromDisk(std::vector<uint8_t>& block, const FlatFilePos& pos, c
     return true;
 }
 
-bool ReadRawBlockFromDisk(std::vector<uint8_t>& block, const CBlockIndex* pindex, const CMessageHeader::MessageStartChars& message_start)
-{
-    const FlatFilePos block_pos{WITH_LOCK(cs_main, return pindex->GetBlockPos())};
-    return ReadRawBlockFromDisk(block, block_pos, message_start);
-}
-
 /** Store block on disk. If dbp is non-nullptr, the file is known to already reside on disk */
 FlatFilePos SaveBlockToDisk(const CBlock& block, int nHeight, CChain& active_chain, const CChainParams& chainparams, const FlatFilePos* dbp)
 {
diff --git a/src/node/blockstorage.h b/src/node/blockstorage.h
index 93e3b1e93..6872848f2 100644
--- a/src/node/blockstorage.h
+++ b/src/node/blockstorage.h
@@ -73,7 +73,6 @@ void UnlinkPrunedFiles(const std::set<int>& setFilesToPrune);
 bool ReadBlockFromDisk(CBlock& block, const FlatFilePos& pos, const Consensus::Params& consensusParams);
 bool ReadBlockFromDisk(CBlock& block, const CBlockIndex* pindex, const Consensus::Params& consensusParams);
 bool ReadRawBlockFromDisk(std::vector<uint8_t>& block, const FlatFilePos& pos, const CMessageHeader::MessageStartChars& message_start);
-bool ReadRawBlockFromDisk(std::vector<uint8_t>& block, const CBlockIndex* pindex, const CMessageHeader::MessageStartChars& message_start);
 
 bool UndoReadFromDisk(CBlockUndo& blockundo, const CBlockIndex* pindex);
 bool WriteUndoDataForBlock(const CBlockUndo& blockundo, BlockValidationState& state, CBlockIndex* pindex, const CChainParams& chainparams) EXCLUSIVE_LOCKS_REQUIRED(cs_main);
```
?

jonatack replied:
Done

hebasto commented on src/chain.h:
```diff
@@ -37,6 +38,8 @@ static constexpr int64_t TIMESTAMP_WINDOW = MAX_FUTURE_BLOCK_TIME;
  */
 static constexpr int64_t MAX_BLOCK_TIME_GAP = 90 * 60;
 
+extern RecursiveMutex cs_main;
```
style nit, here and after: For new code maybe be explicit about the global namespace:
```suggestion
extern RecursiveMutex ::cs_main;
```
?

jonatack replied:
Done (I think).

hebasto commented on src/node/blockstorage.h:
```diff
@@ -73,7 +76,7 @@ bool ReadRawBlockFromDisk(std::vector<uint8_t>& block, const FlatFilePos& pos, c
 
 bool UndoReadFromDisk(CBlockUndo& blockundo, const CBlockIndex* pindex);
-bool WriteUndoDataForBlock(const CBlockUndo& blockundo, BlockValidationState& state, CBlockIndex* pindex, const CChainParams& chainparams);
+bool WriteUndoDataForBlock(const CBlockUndo& blockundo, BlockValidationState& state, CBlockIndex* pindex, const CChainParams& chainparams) EXCLUSIVE_LOCKS_REQUIRED(cs_main);
```
style nit: Is it more readable:
```suggestion
bool WriteUndoDataForBlock(const CBlockUndo& blockundo, BlockValidationState& state, CBlockIndex* pindex, const CChainParams& chainparams)
    EXCLUSIVE_LOCKS_REQUIRED(cs_main);
```
?

jonatack replied:
Not to me ;) but ok, done

jonatack commented:
> While you're making changes, can you also update the commits to drop the `consensus:` prefix, as I don't think that's correct for these changes.

That prefix was used for the commits that update `chain.h`, but sure, will omit the prefix.

hebasto committed:
Remove/inline ReadRawBlockFromDisk(block_data, pindex, message_start)

jonatack commented:
Updated with feedback by @fanquake and @hebasto (thanks!)

hebasto commented on src/index/txindex.cpp:
```diff
@@ -57,7 +57,10 @@ bool TxIndex::WriteBlock(const CBlock& block, const CBlockIndex* pindex)
-    CDiskTxPos pos(pindex->GetBlockPos(), GetSizeOfCompactSize(block.vtx.size()));
+    CDiskTxPos pos{
+        WITH_LOCK(::cs_main, return pindex->GetBlockPos()),
+        GetSizeOfCompactSize(block.vtx.size())
+    };
```
0bfa32b131ed2da8799e1d1239bf390dd612b7bb
style nit: [`clang-format-diff.py`](https://github.com/bitcoin/bitcoin/tree/master/contrib/devtools#clang-format-diffpy) insists on:
```suggestion
        GetSizeOfCompactSize(block.vtx.size())};
```

jonatack replied:
> The `WriteUndoDataForBlock` contains disk operations. It seems suboptimal to hold `::cs_main` for the whole function. 

The lock is already held by the function's caller (and is also needed further down in the function for `pindex->nStatus |= BLOCK_HAVE_UNDO;`), so it seemed best to verify the lock is already held rather than taking it twice unnecessarily. WDYT? I updated the commit description to make this more slightly more clear.

hebasto commented on src/wallet/test/wallet_tests.cpp:
```diff
@@ -137,10 +140,10 @@ BOOST_FIXTURE_TEST_CASE(scan_for_wallet_transactions, TestChain100Setup)
         Assert(m_node.chainman)->m_blockman.PruneOneBlockFile(oldTip->GetBlockPos().nFile);
+        UnlinkPrunedFiles({oldTip->GetBlockPos().nFile});
     }
-    UnlinkPrunedFiles({oldTip->GetBlockPos().nFile});
```
0bfa32b131ed2da8799e1d1239bf390dd612b7bb
`UnlinkPrunedFiles` includes a logging operation. Could we keep logging out of the `::cs_main` lock?

jonatack replied:
done

hebasto reviewed (commented):
The `WriteUndoDataForBlock` contains disk operations. It seems suboptimal to hold `::cs_main` for the whole function. Maybe drop 44037646c8b5e0ee9dcd31094eb6d0a8aefc75cd, and enhance b41973ddc239e47c33562639b9d51a4daf5fef19 with
```diff
--- a/src/node/blockstorage.cpp
+++ b/src/node/blockstorage.cpp
@@ -339,7 +339,7 @@ static bool WriteBlockToDisk(const CBlock& block, FlatFilePos& pos, const CMessa
 bool WriteUndoDataForBlock(const CBlockUndo& blockundo, BlockValidationState& state, CBlockIndex* pindex, const CChainParams& chainparams)
 {
     // Write undo information to disk
-    if (pindex->GetUndoPos().IsNull()) {
+    if (WITH_LOCK(::cs_main, return pindex->GetUndoPos()).IsNull()) {
         FlatFilePos _pos;
         if (!FindUndoPos(state, pindex->nFile, _pos, ::GetSerializeSize(blockundo, CLIENT_VERSION) + 40)) {
             return error("ConnectBlock(): FindUndoPos failed");
```
?

jonatack commented:
Rebased.

Thanks for reviewing @hebasto!

> The `WriteUndoDataForBlock` contains disk operations. It seems suboptimal to hold `::cs_main` for the whole function. 

The lock is already held by the function's caller (and is also needed further down in the function for `pindex->nStatus`) so it seemed best to verify the lock is already held rather than taking it twice unnecessarily. I updated the commit description to make this clearer.

hebasto commented on src/txdb.cpp:
```diff
@@ -308,6 +308,7 @@ bool CBlockTreeDB::LoadBlockIndexGuts(const Consensus::Params& consensusParams,
             CDiskBlockIndex diskindex;
             if (pcursor->GetValue(diskindex)) {
                 // Construct block index object
+                LOCK(::cs_main);
```
Why a scope of this lock is so wide?

jonatack replied:
Scope narrowed per `git diff 71bfaa1 e9560c6`
```diff
--- a/src/txdb.cpp
+++ b/src/txdb.cpp
@@ -308,7 +308,6 @@ bool CBlockTreeDB::LoadBlockIndexGuts(const Consensus::Params& consensusParams,
             CDiskBlockIndex diskindex;
             if (pcursor->GetValue(diskindex)) {
                 // Construct block index object
-                LOCK(::cs_main);
                 CBlockIndex* pindexNew = insertBlockIndex(diskindex.GetBlockHash());
                 pindexNew->pprev          = insertBlockIndex(diskindex.hashPrev);
                 pindexNew->nHeight        = diskindex.nHeight;
@@ -320,11 +319,15 @@ bool CBlockTreeDB::LoadBlockIndexGuts(const Consensus::Params& consensusParams,
                 pindexNew->nTime          = diskindex.nTime;
                 pindexNew->nBits          = diskindex.nBits;
                 pindexNew->nNonce         = diskindex.nNonce;
-                pindexNew->nStatus        = diskindex.nStatus;
                 pindexNew->nTx            = diskindex.nTx;
+                {
+                    LOCK(::cs_main);
+                    pindexNew->nStatus = diskindex.nStatus;
+                }
```

hebasto reviewed (commented):
Approach ACK 71bfaa1c367ea15d73a027dad1fe4601df0c73d2.

achow101 commented:
ACK 71bfaa1c367ea15d73a027dad1fe4601df0c73d2

jonatack commented:
Thank you @hebasto and @achow101 for the reviews; updated with @hebasto's feedback per `git diff 71bfaa1 e9560c6`

hebasto reviewed (approved):
ACK e9560c6a448b4bcf144b0c6fa81ce59ef29863ee

achow101 commented:
ACK e9560c6a448b4bcf144b0c6fa81ce59ef29863ee

vasild commented on src/chain.h:
```diff
@@ -308,20 +313,27 @@ class CBlockIndex
     //! @returns true if the block is assumed-valid; this means it is queued to be
     //!   validated by a background chainstate.
-    bool IsAssumedValid() const { return nStatus & BLOCK_ASSUMED_VALID; }
+    bool IsAssumedValid() const
+    {
+        LOCK(::cs_main);
```
Same as `IsValid()` (except some tests):

<details>
<summary>suggested change</summary>

```diff
diff --git i/src/chain.h w/src/chain.h
index 2a7e420c31..291e684eef 100644
--- i/src/chain.h
+++ w/src/chain.h

[...]
     //! @returns true if the block is assumed-valid; this means it is queued to be
     //!   validated by a background chainstate.
-    bool IsAssumedValid() const
+    bool IsAssumedValid() const EXCLUSIVE_LOCKS_REQUIRED(cs_main)
     {
-        LOCK(::cs_main);
+        AssertLockHeld(cs_main);
         return nStatus & BLOCK_ASSUMED_VALID;
     }
[...]

diff --git i/src/test/validation_chainstatemanager_tests.cpp w/src/test/validation_chainstatemanager_tests.cpp
index 47d9a77126..e3fd1f203a 100644
--- i/src/test/validation_chainstatemanager_tests.cpp
+++ w/src/test/validation_chainstatemanager_tests.cpp
@@ -230,13 +230,16 @@ BOOST_FIXTURE_TEST_CASE(chainstatemanager_activate_snapshot, TestChain100Setup)
     BOOST_CHECK(!chainman.ActiveChainstate().m_from_snapshot_blockhash->IsNull());
     BOOST_CHECK_EQUAL(
         *chainman.ActiveChainstate().m_from_snapshot_blockhash,
         *chainman.SnapshotBlockhash());
 
     // Ensure that the genesis block was not marked assumed-valid.
-    BOOST_CHECK(!chainman.ActiveChain().Genesis()->IsAssumedValid());
+    {
+        LOCK(cs_main);
+        BOOST_CHECK(!chainman.ActiveChain().Genesis()->IsAssumedValid());
+    }
 
     const AssumeutxoData& au_data = *ExpectedAssumeutxo(snapshot_height, ::Params());
     const CBlockIndex* tip = chainman.ActiveTip();
 
     BOOST_CHECK_EQUAL(tip->nChainTx, au_data.nChainTx);
 
@@ -351,16 +354,16 @@ BOOST_FIXTURE_TEST_CASE(chainstatemanager_loadblockindex, TestChain100Setup)
     // tip candidates.
     reload_all_block_indexes();
     BOOST_CHECK_EQUAL(cs1.setBlockIndexCandidates.size(), cs1.m_chain.Height() + 1);
 
     // Mark some region of the chain assumed-valid.
     for (int i = 0; i <= cs1.m_chain.Height(); ++i) {
+        LOCK(::cs_main);
         auto index = cs1.m_chain[i];
 
         if (i < last_assumed_valid_idx && i >= assumed_valid_start_idx) {
-            LOCK(::cs_main);
             index->nStatus = BlockStatus::BLOCK_VALID_TREE | BlockStatus::BLOCK_ASSUMED_VALID;
         }
 
         ++num_indexes;
         if (index->IsAssumedValid()) ++num_assumed_valid;
```
</details>

jonatack replied:
ACK, tested, done.

vasild commented on src/chain.h:
```diff
@@ -308,20 +313,27 @@ class CBlockIndex
+        return nStatus & BLOCK_ASSUMED_VALID;
+    }
 
     //! Raise the validity level of this block index entry.
     //! Returns true if the validity was changed.
     bool RaiseValidity(enum BlockStatus nUpTo)
     {
         assert(!(nUpTo & ~BLOCK_VALID_MASK)); // Only validity flags allowed.
+        LOCK(::cs_main);
```
Same as `IsValid()` (except some tests):


<details>
<summary>suggested change</summary>

```diff
diff --git i/src/chain.h w/src/chain.h
index 2a7e420c31..291e684eef 100644
--- i/src/chain.h
+++ w/src/chain.h

[...]
     //! Raise the validity level of this block index entry.
     //! Returns true if the validity was changed.
-    bool RaiseValidity(enum BlockStatus nUpTo)
+    bool RaiseValidity(enum BlockStatus nUpTo) EXCLUSIVE_LOCKS_REQUIRED(cs_main)
     {
+        AssertLockHeld(cs_main);
         assert(!(nUpTo & ~BLOCK_VALID_MASK)); // Only validity flags allowed.
-        LOCK(::cs_main);
         if (nStatus & BLOCK_FAILED_MASK) return false;
[...]

diff --git i/src/test/fuzz/chain.cpp w/src/test/fuzz/chain.cpp
index 4f97e1ebb1..e4fd2b32f6 100644
--- i/src/test/fuzz/chain.cpp
+++ w/src/test/fuzz/chain.cpp
@@ -55,13 +55,13 @@ FUZZ_TARGET(chain)
             BlockStatus::BLOCK_FAILED_MASK,
             BlockStatus::BLOCK_OPT_WITNESS,
         });
         if (block_status & ~BLOCK_VALID_MASK) {
             continue;
         }
-        (void)disk_block_index->RaiseValidity(block_status);
+        WITH_LOCK(cs_main, (void)disk_block_index->RaiseValidity(block_status));
     }
 
     CBlockIndex block_index{block_header};
     block_index.phashBlock = &zero;
     (void)block_index.GetBlockHash();
     (void)block_index.ToString();
diff --git i/src/test/validation_chainstatemanager_tests.cpp w/src/test/validation_chainstatemanager_tests.cpp
index 47d9a77126..e3fd1f203a 100644
--- i/src/test/validation_chainstatemanager_tests.cpp
+++ w/src/test/validation_chainstatemanager_tests.cpp
@@ -230,13 +230,16 @@ BOOST_FIXTURE_TEST_CASE(chainstatemanager_activate_snapshot, TestChain100Setup)
     BOOST_CHECK(!chainman.ActiveChainstate().m_from_snapshot_blockhash->IsNull());
     BOOST_CHECK_EQUAL(
         *chainman.ActiveChainstate().m_from_snapshot_blockhash,
         *chainman.SnapshotBlockhash());
 
     // Ensure that the genesis block was not marked assumed-valid.
-    BOOST_CHECK(!chainman.ActiveChain().Genesis()->IsAssumedValid());
+    {
+        LOCK(cs_main);
+        BOOST_CHECK(!chainman.ActiveChain().Genesis()->IsAssumedValid());
+    }
 
     const AssumeutxoData& au_data = *ExpectedAssumeutxo(snapshot_height, ::Params());
     const CBlockIndex* tip = chainman.ActiveTip();
 
     BOOST_CHECK_EQUAL(tip->nChainTx, au_data.nChainTx);
 
@@ -351,16 +354,16 @@ BOOST_FIXTURE_TEST_CASE(chainstatemanager_loadblockindex, TestChain100Setup)
     // tip candidates.
     reload_all_block_indexes();
     BOOST_CHECK_EQUAL(cs1.setBlockIndexCandidates.size(), cs1.m_chain.Height() + 1);
 
     // Mark some region of the chain assumed-valid.
     for (int i = 0; i <= cs1.m_chain.Height(); ++i) {
+        LOCK(::cs_main);
         auto index = cs1.m_chain[i];
 
         if (i < last_assumed_valid_idx && i >= assumed_valid_start_idx) {
-            LOCK(::cs_main);
             index->nStatus = BlockStatus::BLOCK_VALID_TREE | BlockStatus::BLOCK_ASSUMED_VALID;
         }
 
         ++num_indexes;
         if (index->IsAssumedValid()) ++num_assumed_valid;
```
</details>

jonatack replied:
Done.

vasild commented on src/node/blockstorage.cpp:
```diff
@@ -427,6 +427,7 @@ CBlockIndex* BlockManager::GetLastCheckpoint(const CCheckpointData& data)
 
 bool IsBlockPruned(const CBlockIndex* pblockindex)
 {
+    LOCK(::cs_main);
```
Same as `IsValid()`, except 3 of 4 callers of `IsBlockPruned()` already own `cs_main`. The 4th caller, `blockToJSON()` could be changed to acquire the mutex.

<details>
<summary>suggested change</summary>

```diff
diff --git i/src/node/blockstorage.cpp w/src/node/blockstorage.cpp
index 5074080ed1..2217a4d6c8 100644
--- i/src/node/blockstorage.cpp
+++ w/src/node/blockstorage.cpp
@@ -424,13 +424,13 @@ CBlockIndex* BlockManager::GetLastCheckpoint(const CCheckpointData& data)
     }
     return nullptr;
 }
 
 bool IsBlockPruned(const CBlockIndex* pblockindex)
 {
-    LOCK(::cs_main);
+    AssertLockHeld(cs_main);
     return (fHavePruned && !(pblockindex->nStatus & BLOCK_HAVE_DATA) && pblockindex->nTx > 0);
 }
 
 // If we're using -prune with -reindex, then delete block files that will be ignored by the
 // reindex.  Since reindexing works by starting at block file 0 and looping until a blockfile
 // is missing, do the same here to delete any later block files after a gap.  Also delete all
diff --git i/src/node/blockstorage.h w/src/node/blockstorage.h
index 7efa7e3b72..0cdab404da 100644
--- i/src/node/blockstorage.h
+++ w/src/node/blockstorage.h
@@ -163,13 +163,13 @@ public:
     {
         Unload();
     }
 };
 
 //! Check whether the block associated with this index entry is pruned or not.
-bool IsBlockPruned(const CBlockIndex* pblockindex);
+bool IsBlockPruned(const CBlockIndex* pblockindex) EXCLUSIVE_LOCKS_REQUIRED(cs_main);
 
 void CleanupBlockRevFiles();
 
 /** Open a block file (blk?????.dat) */
 FILE* OpenBlockFile(const FlatFilePos& pos, bool fReadOnly = false);
 /** Translation to a filesystem path */
diff --git i/src/rpc/blockchain.cpp w/src/rpc/blockchain.cpp
index 455c32c57a..d7c1572099 100644
--- i/src/rpc/blockchain.cpp
+++ w/src/rpc/blockchain.cpp
@@ -172,13 +172,17 @@ UniValue blockToJSON(const CBlock& block, const CBlockIndex* tip, const CBlockIn
             }
             break;
 
         case TxVerbosity::SHOW_DETAILS:
         case TxVerbosity::SHOW_DETAILS_AND_PREVOUT:
             CBlockUndo blockUndo;
-            const bool have_undo = !IsBlockPruned(blockindex) && UndoReadFromDisk(blockUndo, blockindex);
+            bool have_undo;
+            {
+                LOCK(cs_main);
+                have_undo = !IsBlockPruned(blockindex) && UndoReadFromDisk(blockUndo, blockindex);
+            }
 
             for (size_t i = 0; i < block.vtx.size(); ++i) {
                 const CTransactionRef& tx = block.vtx.at(i);
                 // coinbase transaction (i.e. i == 0) doesn't have undo data
                 const CTxUndo* txundo = (have_undo && i > 0) ? &blockUndo.vtxundo.at(i - 1) : nullptr;
                 UniValue objTx(UniValue::VOBJ);
@@ -925,14 +929,15 @@ static RPCHelpMan getblockheader()
 
     return blockheaderToJSON(tip, pblockindex);
 },
     };
 }
 
-static CBlock GetBlockChecked(const CBlockIndex* pblockindex)
+static CBlock GetBlockChecked(const CBlockIndex* pblockindex) EXCLUSIVE_LOCKS_REQUIRED(cs_main)
 {
+    AssertLockHeld(cs_main);
     CBlock block;
     if (IsBlockPruned(pblockindex)) {
         throw JSONRPCError(RPC_MISC_ERROR, "Block not available (pruned data)");
     }
 
     if (!ReadBlockFromDisk(block, pblockindex, Params().GetConsensus())) {
@@ -942,14 +947,15 @@ static CBlock GetBlockChecked(const CBlockIndex* pblockindex)
         throw JSONRPCError(RPC_MISC_ERROR, "Block not found on disk");
     }
 
     return block;
 }
 
-static CBlockUndo GetUndoChecked(const CBlockIndex* pblockindex)
+static CBlockUndo GetUndoChecked(const CBlockIndex* pblockindex) EXCLUSIVE_LOCKS_REQUIRED(cs_main)
 {
+    AssertLockHeld(cs_main);
     CBlockUndo blockUndo;
     if (IsBlockPruned(pblockindex)) {
         throw JSONRPCError(RPC_MISC_ERROR, "Undo data not available (pruned data)");
     }
 
     if (!UndoReadFromDisk(blockUndo, pblockindex)) {
```
</details>

I think it is better to call `UndoReadFromDisk()` also under the same acquisition of `cs_main` as `IsBlockPruned()` (as in the patch above) because that would ensure `nStatus` remains unchanged until after the disk read has completed (i.e. we don't end up with a status that indicates no file on disk and still trying to read the file, if `nStatus` is changed after we read it and before we read the file from disk).

jonatack replied:
I had planned to add lock annotations to `IsBlockPruned()` in a follow-up, but I may as well do it here with the other changes. Done. 

If this pull is now too large to attract re-ACKs, in that case it could be reverted to its original version with the first 3 commits only, and the `nStatus` preparation and change could be proposed separately. We'll see.

vasild commented on src/chain.h:
```diff
@@ -223,8 +226,9 @@ class CBlockIndex
-    FlatFilePos GetBlockPos() const
+    FlatFilePos GetBlockPos() const EXCLUSIVE_LOCKS_REQUIRED(::cs_main)
     {
+        AssertLockHeld(::cs_main);
         FlatFilePos ret;
         if (nStatus & BLOCK_HAVE_DATA) {
             ret.nFile = nFile;
```
This access pattern requires that `nFile` stays in sync with `nStatus` - we don't want `nStatus` to be changed concurrently by another thread after we have read it and entered this `if` but before we have read `nFile`. That is to say - `nFile`, `nDataPos` and `nUndoPos` should be protected by the same mechanism as `nStatus`. This is easy:


<details>
<summary>suggested change</summary>

```diff
diff --git i/src/chain.h w/src/chain.h
index 2a7e420c31..d8f108b466 100644
--- i/src/chain.h
+++ w/src/chain.h
@@ -161,19 +161,19 @@ public:
     CBlockIndex* pskip{nullptr};
 
     //! height of the entry in the chain. The genesis block has height 0
     int nHeight{0};
 
     //! Which # file this block is stored in (blk?????.dat)
-    int nFile{0};
+    int nFile GUARDED_BY(::cs_main){0};
 
     //! Byte offset within blk?????.dat where this block's data is stored
-    unsigned int nDataPos{0};
+    unsigned int nDataPos GUARDED_BY(::cs_main){0};
 
     //! Byte offset within rev?????.dat where this block's undo data is stored
-    unsigned int nUndoPos{0};
+    unsigned int nUndoPos GUARDED_BY(::cs_main){0};
 
     //! (memory only) Total amount of work (expected number of hashes) in the chain up to and including this block
     arith_uint256 nChainWork{};
 
     //! Number of transactions in this block.
     //! Note: in a potential headers-first mode, this number cannot be relied upon
diff --git i/src/txdb.cpp w/src/txdb.cpp
index 3fd5ca5b19..e2aed1da7c 100644
--- i/src/txdb.cpp
+++ w/src/txdb.cpp
@@ -308,23 +308,23 @@ bool CBlockTreeDB::LoadBlockIndexGuts(const Consensus::Params& consensusParams,
             CDiskBlockIndex diskindex;
             if (pcursor->GetValue(diskindex)) {
                 // Construct block index object
                 CBlockIndex* pindexNew = insertBlockIndex(diskindex.GetBlockHash());
                 pindexNew->pprev          = insertBlockIndex(diskindex.hashPrev);
                 pindexNew->nHeight        = diskindex.nHeight;
-                pindexNew->nFile          = diskindex.nFile;
-                pindexNew->nDataPos       = diskindex.nDataPos;
-                pindexNew->nUndoPos       = diskindex.nUndoPos;
                 pindexNew->nVersion       = diskindex.nVersion;
                 pindexNew->hashMerkleRoot = diskindex.hashMerkleRoot;
                 pindexNew->nTime          = diskindex.nTime;
                 pindexNew->nBits          = diskindex.nBits;
                 pindexNew->nNonce         = diskindex.nNonce;
                 pindexNew->nTx            = diskindex.nTx;
                 {
                     LOCK(::cs_main);
+                    pindexNew->nFile = diskindex.nFile;
+                    pindexNew->nDataPos = diskindex.nDataPos;
+                    pindexNew->nUndoPos = diskindex.nUndoPos;
                     pindexNew->nStatus = diskindex.nStatus;
                 }
 
                 if (!CheckProofOfWork(pindexNew->GetBlockHash(), pindexNew->nBits, consensusParams)) {
                     return error("%s: CheckProofOfWork failed: %s", __func__, pindexNew->ToString());
                 }
```
</details>

jonatack replied:
Done in the `nStatus` commit (thanks!)

vasild reviewed (commented):
Approach ACK e9560c6a448b4bcf144b0c6fa81ce59ef29863ee

I think this does not fully resolve https://github.com/bitcoin/bitcoin/issues/17161 but is an improvement that can go in as is. I think the issue is not fully resolved because we can still have the `nStatus` variable go out of sync with the disk contents - we can read `nStatus & BLOCK_HAVE_DATA` to `true` under `cs_main`, but immediately after the read, before `cs_main` is released, prune may delete the files from disk:

https://github.com/bitcoin/bitcoin/blob/d0bf9bb6a539f151ec92725d20a2b6c22cb095a5/src/validation.cpp#L2296

jonatack committed:
Require IsBlockPruned() to hold mutex cs_main

Co-authored-by: Vasil Dimov <vd@FreeBSD.org>

jonatack committed:
Guard CBlockIndex::nStatus/nFile/nDataPos/nUndoPos by cs_main

Co-authored-by: Vasil Dimov <vd@FreeBSD.org>

vasild committed:
Require CBlockIndex::IsAssumedValid() to hold cs_main

vasild committed:
Require CBlockIndex::RaiseValidity() to hold cs_main

vasild committed:
Require CBlockIndex::IsValid() to hold cs_main

jonatack commented:
Addressed review feedback by @vasild and updated the pull title/description. Invalidates previous ACKs by @hebasto, @achow101 and @vasild.

vasild commented on src/test/util/blockfilter.cpp:
```diff
@@ -11,6 +11,8 @@
 
 bool ComputeFilter(BlockFilterType filter_type, const CBlockIndex* block_index, BlockFilter& filter)
 {
+    LOCK(::cs_main);
```
I am just curious - why prefix with `::`? There is no conflict with a local symbol under the same name, so the `::` shouldn't be necessary? Right now in the codebase there are 154 occurrences of `::cs_main` and 421 occurrences of `cs_main` (without the prefix).

hebasto replied:
I do prefer specify the global namespace for the sake being explicit.

jonatack replied:
Right, my understanding from review feedback is that it is preferred for new or updated code.

vasild replied:
For all global variables or just for `cs_main`? Global variables are already supposed to start with `g_`, so prefixing them with `::` is kind of redundant if the purpose is to show to the reader that this is a global variable.

MarcoFalke replied:
`cs_main` doesn't start with `g_`

vasild reviewed (approved):
ACK cf2145beb26f0cf3016709b2159bcb4053b58eb9

jonatack commented:
Rebased.

This has concept ACKs by @MarcoFalke and @promag and ACKs by @hebasto, @achow101 and @vasild.

Review/re-ACKs welcome.

jamesob commented:
Benched 6ea56827842b9b2bd730edc38f3a7b1f46f6247b; no performance impact :+1:.

|          bench name           |                                                                                                     command                                                                                                      |
|-------------------------------|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ibd.local.range.500000.540000 | `bitcoind -dbcache=10000 -debug=coindb -debug=bench -listen=0 -connect=0 -addnode=127.0.0.1:8888 -prune=9999999 -printtoconsole=0 -assumevalid=000000000000000000176c192f42ad13ab159fdb20198b87e7ba3c001e47b876` |


### #22932 vs. $mergebase (absolute)
|                  bench name                   |  x  |           #22932           |         $mergebase         |
|-----------------------------------------------|----:|----------------------------|----------------------------|
| ibd.local.range.500000.540000.total_secs      |   2 | 3093.6633 (± 17.1804)      | 3102.2931 (± 0.4422)       |
| ibd.local.range.500000.540000.peak_rss_KiB    |   2 | 6355854.0000 (± 1646.0000) | 6355764.0000 (± 1492.0000) |
| ibd.local.range.500000.540000.cpu_kernel_secs |   2 | 263.2300 (± 3.3200)        | 264.0900 (± 0.1700)        |
| ibd.local.range.500000.540000.cpu_user_secs   |   2 | 13020.2400 (± 4.7400)      | 13015.7250 (± 8.2850)      |


### #22932 vs. $mergebase (relative)
|                  bench name                   |  x  | #22932 | $mergebase |
|-----------------------------------------------|----:|-------:|-----------:|
| ibd.local.range.500000.540000.total_secs      |   2 |  1.000 |      1.003 |
| ibd.local.range.500000.540000.peak_rss_KiB    |   2 |  1.000 |      1.000 |
| ibd.local.range.500000.540000.cpu_kernel_secs |   2 |  1.000 |      1.003 |
| ibd.local.range.500000.540000.cpu_user_secs   |   2 |  1.000 |      1.000 |

laanwj commented:
Code review ACK 6ea56827842b9b2bd730edc38f3a7b1f46f6247b

vasild reviewed (commented):
ACK 6ea56827842b9b2bd730edc38f3a7b1f46f6247b

MarcoFalke commented on src/chain.h:
```diff
@@ -382,6 +382,7 @@ class CDiskBlockIndex : public CBlockIndex
 
     SERIALIZE_METHODS(CDiskBlockIndex, obj)
     {
+        LOCK(::cs_main);
```
can you explain why this is needed?

`CDiskBlockIndex` creates a copy of the data and it isn't used in validation or blockstorage code. It is only used to serialize a struct and then discarded.

jonatack replied:
Having a look.

jonatack replied:
> can you explain why this is needed?
> 
> `CDiskBlockIndex` creates a copy of the data and it isn't used in validation or blockstorage code. It is only used to serialize a struct and then discarded.

Good point. Opened #24199 as an approach to address this.

MarcoFalke commented on src/txdb.cpp:
```diff
@@ -311,19 +311,23 @@ bool CBlockTreeDB::LoadBlockIndexGuts(const Consensus::Params& consensusParams,
-                pindexNew->nStatus        = diskindex.nStatus;
                 pindexNew->nTx            = diskindex.nTx;
+                {
+                    LOCK(::cs_main);
```
Can you explain why this is needed? The calling function should already hold cs_main?

jonatack replied:
Good catch--we can drop the lock in favor of an annotation; done in #24197.
